### PR TITLE
Factory class for Breit Wheeler and Synchrotron engines

### PR DESCRIPTION
**Pull request by lucafedeli88**
This PR implements  an interface between WarpX and the QED modules (Breit Wheeler and Quantum Synchrotron) in PICSAR, in the form of factory classes able to create functors to be used in particle loops. These functors are not yet used (except for the initialization of the optical depth). This will be the goal of another PR, since the present one is already quite large.

More specifically, the factory classes are implemented in ```QED/BreitWheelerEngineWrapper.[H/cpp]``` and  ```QED/QuantumSyncEngineWrapper.[H/cpp]``` . They are capable of generating the functors, generating the lookup tables required by the engines (if WarpX is compiled with  QED_TABLE_GEN=TRUE ), exporting and importing the lookup tables to and from binary raw data (```Vector<char>```).  Enabling table generation (```QED_TABLE_GEN=TRUE```) requires up-to-date Boost libraries.  

The  ```QED``` folder contains the following files:
-  ```BreitWheelerEngineInnards.H``` a struct to store control parameters and lookup tables data
- ```QuantumSyncEngineInnards.H``` same as above for QS
- ```BreitWheelerEngineTableBuilder.[H/cpp]``` a class encapsulating table generation capability
- ```QuantumSyncEngineTableBuilder.[H/cpp]``` same as above for QS
- ```BreitWheelerEngineWrapper.[H/cpp]``` the factory class (functors are defined in .H)
- ```QuantumSyncEngineWrapper.[H/cpp]``` same as above for QS
- ```QedChiFunctions.H``` wrappers for functions to calculate the _chi_ parameter
- ```QedTableParserHelperFunctions.H``` helper functions to operate safely on raw binary data
- ```QedWrapperCommons.H``` a common header with few definitions.

```MultiParticleContainer``` has been modified to check if any species needs either the QS or the BW engine. If this is the case, it initializes the required engine(s) and it provides a smart pointer to the species. 
Table generation/loading is done reading parameters in the inputfile (example inputfiles are provided).

For test purposes, table loading/generation can be bypassed with ``` qed_qs.ignore_tables_for_test = 1 ``` or  ``` qed_bw.ignore_tables_for_test = 1 ``` in the inputfile. 

**Warning: some C++14 features are used (in CPU code). Is this OK?** 
**Warning: picsar should be on the QED branch to compile with QED=TRUE** (this will be fixed soon)
**Warning:  mini-PR https://github.com/ECP-WarpX/WarpX/pull/491 (bugfixing) has already been merged with this branch**

### CODE REVIEW

**Comment by lucafedeli88** [comment]
I put [WIP] back because I've discovered an issue while doing some tests on GPU

**Comment by MaxThevenet** [comment]
Thanks for that, I'll wait for your signal before reviewing, then.

**Comment by lucafedeli88** [comment]
I've removed [WIP] from the title because commit 57080a6  solved the issue that I was experiencing. I've also moved two example inputfiles out of the Example folder since they require to compile WarpX with 
```QED=TRUE QED_TABLE_GEN=TRUE``` (i.e. Boost library is required)

**Comment by MaxThevenet** [comment]
@lucafedeli88 I see you added `[WIP]` back, should I wait before reviewing it?

**Comment by MaxThevenet** [comment]
Oops sorry, the name actually not changed, you just pushed one commit.
Anyway, should I wait before reviewing it?

**Comment by lucafedeli88** [comment]
> Oops sorry, the name actually did not changed, you just pushed one commit.
> Anyway, should I wait before reviewing it?

It's just that I've found a bug while developing https://github.com/ECP-WarpX/WarpX/pull/482 . However, It made more sense to correct it here. It involved just a couple of lines anyway, so I think that you can review it.

**Comment by MaxThevenet** [comment]
cool, thanks.

File: Source/QED/BreitWheelerEngineWrapper.cpp [hunk 1]
```diff
@@ -1,16 +1,205 @@
...
+
+    //Data
+    Vector<Real> tndt_coords(innards.ctrl.chi_phot_tdndt_how_many);
+    Vector<Real> tndt_data(innards.ctrl.chi_phot_tdndt_how_many);
+    Vector<Real> cum_tab_coords1(innards.ctrl.chi_phot_tpair_how_many);
+    Vector<Real> cum_tab_coords2(innards.ctrl.chi_frac_tpair_how_many);
+    Vector<Real> cum_tab_data(innards.ctrl.chi_phot_tpair_how_many*
+        innards.ctrl.chi_frac_tpair_how_many);
+
+    tie(is_ok, tndt_coords, p_data) =
```
Comment (MaxThevenet) [inline Source/QED/BreitWheelerEngineWrapper.cpp]: Could you add a small comment before this line, to explain what this block is doing? Same thing for the few similar blocks below (I am just not sure of what `tndt_coords`, `tndt_data`, `cum_tab_coords1` `cum_tab_coords2` stand for), and for the QSR module. Some description could go to the `innard` files.

File: Source/QED/BreitWheelerEngineInnards.H [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    // Control parameters (a POD struct)
+    picsar::multi_physics::breit_wheeler_engine_ctrl<amrex::Real> ctrl;
+
+    //Lookup table data (should stay on GPU)
+    amrex::Gpu::ManagedVector<amrex::Real> TTfunc_coords;
+    amrex::Gpu::ManagedVector<amrex::Real> TTfunc_data;
+
+    amrex::Gpu::ManagedVector<amrex::Real> cum_distrib_coords_1;
+    amrex::Gpu::ManagedVector<amrex::Real> cum_distrib_coords_2;
+    amrex::Gpu::ManagedVector<amrex::Real> cum_distrib_data;
```
Comment (MaxThevenet) [inline Source/QED/BreitWheelerEngineInnards.H]: Could you add a small description for each member variable?

File: Source/QED/BreitWheelerEngineWrapper.cpp [hunk 3]
```diff
@@ -1,16 +1,205 @@
...
+    add_data_to_buf(innards.ctrl.chi_phot_tpair_max);
+    add_data_to_buf(innards.ctrl.chi_phot_tpair_how_many);
+    add_data_to_buf(innards.ctrl.chi_frac_tpair_how_many);
+
+    auto add_data_to_buf_vv = [&res](const auto* data, size_t how_many){
+        res.insert(res.end(),
+        reinterpret_cast<const char*>(data),
+        reinterpret_cast<const char*>(data) +
+        sizeof(*data)*how_many);
+    };
```
Comment (MaxThevenet) [inline Source/QED/BreitWheelerEngineWrapper.cpp]: Couldn't `add_data_to_buf_vv` be used above too, passing `1` as argument `how_many`? Or is there another difference (for instance is the `const` qualifier for first argument necessary in one case and not the other)?

File: Source/QED/QuantumSyncEngineWrapper.H [hunk 4]
```diff
@@ -22,33 +42,236 @@ using WarpXQuantumSynchrotronWrapper =
...
+     */
+    template <size_t sampling>
+    AMREX_GPU_DEVICE
+    AMREX_FORCE_INLINE
+    void operator()(
+    amrex::Real* px, amrex::Real* py, amrex::Real* pz,
+    amrex::Real ex, amrex::Real ey, amrex::Real ez,
+    amrex::Real bx, amrex::Real by, amrex::Real bz,
+    amrex::Real weight,
+    amrex::Real* g_px, amrex::Real* g_py, amrex::Real* g_pz,
```
Comment (MaxThevenet) [inline Source/QED/QuantumSyncEngineWrapper.H]: Same comment as for `BreitWheeler`, could some of these be in the constructor?

File: Source/QED/QuantumSyncEngineWrapper.H [hunk 5]
```diff
@@ -22,33 +42,236 @@ using WarpXQuantumSynchrotronWrapper =
...
+            ctrl,
+            rand_zero_one_minus_epsi.data());
+    }
+
+private:
+    //laser wavelenght is not used with SI units
+    const amrex::Real dummy_lambda{1.0};
+
+    const PicsarQuantumSynchrotronCtrl ctrl;
+    const picsar::multi_physics::lookup_2d<amrex::Real> lookup_table;
```
Comment (MaxThevenet) [inline Source/QED/QuantumSyncEngineWrapper.H]: Same comment as for `BreitWheeler`, should this be e.g. a pointer?

File: Source/QED/QuantumSyncEngineWrapper.cpp [hunk 6]
```diff
@@ -14,4 +21,181 @@ QuantumSynchrotronEngine::build_optical_depth_functor ()
...
+    add_data_to_buf(innards.ctrl.chi_part_tem_max);
+    add_data_to_buf(innards.ctrl.chi_part_tem_how_many);
+    add_data_to_buf(innards.ctrl.prob_tem_how_many);
+
+    auto add_data_to_buf_vv = [&res](const auto* data, size_t how_many){
+        res.insert(res.end(),
+        reinterpret_cast<const char*>(data),
+        reinterpret_cast<const char*>(data) +
+        sizeof(*data)*how_many);
+    };
```
Comment (MaxThevenet) [inline Source/QED/QuantumSyncEngineWrapper.cpp]: Actually could this be in, e.g.,  `QedWrapperCommons.H`? Just a suggestion.

**Comment by MaxThevenet** [comment]
Oh, and could you fix conflicts with `dev` branch?

File: Examples/Modules/qed/breit_wheeler/inputs.2d_test_tau_init [hunk 7]
```diff
@@ -64,3 +64,7 @@ photons.uz_th = 100.
 photons.do_qed = 1
 photons.do_qed_breit_wheeler = 1
 #################################
+
+##########QED TABLES#############
+qed_bw.ignore_tables_for_test = 1
+#################################
```
[the remark this answers is not in the document]
Comment (lucafedeli88) [reply Examples/Modules/qed/breit_wheeler/inputs.2d_test_tau_init]: done

File: Examples/Modules/qed/quantum_synchrotron/inputs.2d_test_tau_init [hunk 8]
```diff
@@ -88,3 +88,7 @@ positrons.uz_th = 100.
 positrons.do_qed = 1
 positrons.do_qed_quantum_sync = 1
 #################################
+
+##########QED TABLES#############
+qed_qs.ignore_tables_for_test = 1
+#################################
```
[the remark this answers is not in the document]
Comment (lucafedeli88) [reply Examples/Modules/qed/quantum_synchrotron/inputs.2d_test_tau_init]: sure

File: QedExamplesRequiringBoost/quantum_synchrotron/__needs_boost_inputs.2d_test_table_gen [hunk 9]
```diff
@@ -0,0 +1,110 @@
+#An inputfile to demonstrate lookup table generation for the Quantum Synchrotron engine
```
[the remark this answers is not in the document]
Comment (lucafedeli88) [reply QedExamplesRequiringBoost/quantum_synchrotron/__needs_boost_inputs.2d_test_table_gen]: ok. I agree

[the remark this answers is not in the document]
Comment (lucafedeli88) [reply QedExamplesRequiringBoost/quantum_synchrotron/__needs_boost_inputs.2d_test_table_gen]: done

File: Source/Particles/MultiParticleContainer.H [hunk 10]
```diff
@@ -219,9 +222,44 @@ protected:
...
-    //Initialize QED engines and provides smart pointers
-    //to species who need QED processes
+    /**
+     * Initialize QED engines and provides smart pointers
+     * to species who need QED processes
+     */
     void InitQED ();
+
+    //Flags to store if at least one species needs a QED process
+    bool someone_has_quantum_sync = false;
```
[the remark this answers is not in the document]
Comment (lucafedeli88) [reply Source/Particles/MultiParticleContainer.H]: I agree. 
I've just corrected that.

File: Source/Particles/MultiParticleContainer.cpp [hunk 11]
```diff
@@ -743,11 +751,209 @@ void MultiParticleContainer::InitQED ()
...
+}
+
+void MultiParticleContainer::InitQuantumSync ()
+{
+    bool generate_table;
+    PicsarQuantumSynchrotronCtrl ctrl;
+    std::string filename;
+    std::tie(generate_table, filename, ctrl) = ParseQuantumSyncParams();
+
+    //Only temporary for test purposes, will be removed
```
[the remark this answers is not in the document]
Comment (lucafedeli88) [reply Source/Particles/MultiParticleContainer.cpp]: In general, if QED effects are enabled lookup tables should be either generated (this requires Boost) or loaded from disk. The initialization of the optical depth is the only step which does not require QED tables, so I've introduced this setting to be able to add a test without having the lookup tables. 
We could add some default tables (100s kB binary files) directly in the git repository, but since they are likely to change several times in this phase, I would like to delay that step. I would like to discuss the option of using `git-annex` (this would allow me to change the lookup tables as many times as I want without polluting the git repository). 

So, basically, the answer to your question is: if we want to keep a test of the optical depth initialization, this flag should probably stay there until we find a definitive solution for the lookup tables.

File: Source/Particles/MultiParticleContainer.cpp [hunk 12]
```diff
@@ -743,11 +751,209 @@ void MultiParticleContainer::InitQED ()
...
+    pp.query("ignore_tables_for_test", ignore_tables);
+    if(ignore_tables) return;
+    //_________________________________________________
+
+
+    if(generate_table && ParallelDescriptor::IOProcessor()){
+        shr_p_qs_engine->compute_lookup_tables(ctrl);
+        Vector<char> all_data = shr_p_qs_engine->export_lookup_tables_data();
+        WarpXUtilIO::WriteBinaryDataOnFile(filename, all_data);
+   }
```
[the remark this answers is not in the document]
Comment (lucafedeli88) [reply Source/Particles/MultiParticleContainer.cpp]: I don't have access to Summit :-(  (the only GPU-based supercomputer I've access to is Cooley, whose GPUs are too old for AMReX). I have access to a Linux Workstation  equipped with a NVIDIA Quadro GV100 Volta card. I've compiled the code with `USE_GPU=TRUE` and table generation (on CPU) works.

File: Source/Particles/MultiParticleContainer.H [hunk 13]
```diff
@@ -219,9 +222,44 @@ protected:
...
+    bool someone_has_breit_wheeler = false;
```
[the remark this answers is not in the document]
Comment (lucafedeli88) [reply Source/Particles/MultiParticleContainer.H]: I think that this is a good idea! I've refactored my code accordingly. Is it ok if I don't follow this convention for simple structs (i.e. pure data without methods, like `innards` for the QED engines)?

File: Source/Particles/MultiParticleContainer.cpp [hunk 14]
```diff
@@ -743,11 +751,209 @@ void MultiParticleContainer::InitQED ()
...
+       if(generate_table && ParallelDescriptor::IOProcessor()){
+            amrex::Print() << "Warning, Quantum Synchrotron table will be loaded, not generated. \n";
+       }
+        table_name = load_table_name;
+        generate_table = false;
+    }
+
+#ifndef WARPX_QED_TABLE_GEN
+    if(generate_table){
+            amrex::Error("Error: use QED_TABLE_GEN=TRUE to enable table generation!\n");
```
[the remark this answers is not in the document]
Comment (lucafedeli88) [reply Source/Particles/MultiParticleContainer.cpp]: I agree. I've just done that.

File: Source/Particles/MultiParticleContainer.cpp [hunk 15]
```diff
@@ -743,11 +751,209 @@ void MultiParticleContainer::InitQED ()
...
+            amrex::Print() << "Warning, Breit Wheeler table will be loaded, not generated. \n";
+       }
+        table_name = load_table_name;
+        generate_table = false;
+    }
+
+#ifndef WARPX_QED_TABLE_GEN
+    if(generate_table){
+        if(ParallelDescriptor::IOProcessor()){
+            amrex::Error("Error: use QED_TABLE_GEN=TRUE to enable table generation!\n");
```
[the remark this answers is not in the document]
Comment (lucafedeli88) [reply Source/Particles/MultiParticleContainer.cpp]: I agree. I've just done that.

File: Source/Particles/MultiParticleContainer.cpp [hunk 16]
```diff
@@ -1,14 +1,22 @@
...
+
 #include <limits>
 #include <algorithm>
 #include <string>
-#include <memory>
 
 using namespace amrex;
 
+//literals allow to write "aa"s, which means that "aa" is an std::string (C++14)
+using namespace std::literals;
```
[the remark this answers is not in the document]
Comment (lucafedeli88) [reply Source/Particles/MultiParticleContainer.cpp]: I've changed how I check if a string is empty, so this is no longer relevant (I've removed `std::literals`)... Probably it wasn't really necessary even before.
However, note that I still need C++14, namely for the construct:
```
    auto add_data_to_buf = [&res](auto data){
        res.insert(res.end(),
        reinterpret_cast<const char*>(&data),
        reinterpret_cast<const char*>(&data) +
        sizeof(data));
    };
```
in `BreitWheelerEngineWrapper.cpp` and `QunatumSyncEngineWrapper.cpp`

File: Source/QED/QedChiFunctions.H [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+    */
+    AMREX_GPU_DEVICE
+    AMREX_FORCE_INLINE
+    amrex::Real chi_photon(
+        amrex::Real px, amrex::Real py, amrex::Real pz,
+        amrex::Real ex, amrex::Real ey, amrex::Real ez,
+        amrex::Real bx, amrex::Real by, amrex::Real bz)
+    {
+        //laser wavelength is unused if SI units are set
+        const amrex::Real dummy_lambda = 1.0;
```
[the remark this answers is not in the document]
Comment (lucafedeli88) [reply Source/QED/QedChiFunctions.H]: Only if the variable member is static, apparently ( https://stackoverflow.com/questions/27408819/why-cant-non-static-data-members-be-constexpr )

File: Source/QED/QedChiFunctions.H [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+    */
+    AMREX_GPU_DEVICE
+    AMREX_FORCE_INLINE
+    amrex::Real chi_lepton(
+        amrex::Real px, amrex::Real py, amrex::Real pz,
+        amrex::Real ex, amrex::Real ey, amrex::Real ez,
+        amrex::Real bx, amrex::Real by, amrex::Real bz)
+    {
+        //laser wavelength is unused if SI units are set
+        const amrex::Real dummy_lambda = 1.0;
```
[the remark this answers is not in the document]
Comment (lucafedeli88) [reply Source/QED/QedChiFunctions.H]: see above

File: Source/QED/BreitWheelerEngineWrapper.H [hunk 19]
```diff
@@ -2,53 +2,279 @@
...
+    #include "BreitWheelerEngineTableBuilder.H"
+#endif
+
+#include <string>
 
-using WarpXBreitWheelerWrapper =
-    picsar::multi_physics::breit_wheeler_engine<amrex::Real, DummyStruct>;
+//Some handy aliases
+using PicsarBreitWheelerEngine = picsar::multi_physics::
+    breit_wheeler_engine<amrex::Real, QedUtils::DummyStruct>;
```
[the remark this answers is not in the document]
Comment (lucafedeli88) [reply Source/QED/BreitWheelerEngineWrapper.H]: I've added some comments to explain why I am using an empty data structure there.
I think that having `DummyStruct` in `QedUtils` is the best solution, since it is used twice (in the two engine headers). An alternative could be to define something like `BWDummyStruct` and `QSDummyStruct` in the two headers.

File: Source/QED/BreitWheelerEngineWrapper.H [hunk 20]
```diff
@@ -2,53 +2,279 @@
...
 
+/**
+ * Functor to evolve the optical depth of photons due to the
+ * Breit-Wheeler process
+ */
+class BreitWheelerEvolveOpticalDepth
+{
+public:
+    /**
+     * TO ADD!
```
[the remark this answers is not in the document]
Comment (lucafedeli88) [reply Source/QED/BreitWheelerEngineWrapper.H]: sorry, I've just added the missing comment.

File: Source/QED/BreitWheelerEngineWrapper.H [hunk 21]
```diff
@@ -2,53 +2,279 @@
...
+     * @param[in] px,py,pz momentum components of the photon (SI units)
+     * @param[in] ex,ey,ez electric field components (SI units)
+     * @param[in] bx,by,bz magnetic field components (SI units)
+     * @param[in] dt timestep (SI units)
+     * @param[in,out] opt_depth optical depth of the photon. It is modified by the method.
+     * @return a flag which is true if optical depth becomes negative (i.e. a pair has to be generated).
+     */
+    AMREX_GPU_DEVICE
+    AMREX_FORCE_INLINE
+    bool operator()(
```
[the remark this answers is not in the document]
Comment (lucafedeli88) [reply Source/QED/BreitWheelerEngineWrapper.H]: I've changed that anyway

File: Source/QED/BreitWheelerEngineWrapper.H [hunk 22]
```diff
@@ -2,53 +2,279 @@
...
+    amrex::Real px, amrex::Real py, amrex::Real pz,
+    amrex::Real ex, amrex::Real ey, amrex::Real ez,
+    amrex::Real bx, amrex::Real by, amrex::Real bz,
+    amrex::Real dt, amrex::Real& opt_depth) const
```
[the remark this answers is not in the document]
Comment (lucafedeli88) [reply Source/QED/BreitWheelerEngineWrapper.H]: I've added `noexcept` to `operator ()` for all the functors

File: Source/QED/BreitWheelerEngineWrapper.cpp [hunk 23]
```diff
@@ -1,16 +1,205 @@
...
+bool BreitWheelerEngine::are_lookup_tables_initialized () const
+{
+    return lookup_tables_initialized;
+}
+
+bool
+BreitWheelerEngine::init_lookup_tables_from_raw_data (
+    const Vector<char>& raw_data)
+{
+    const char* p_data = raw_data.data();
```
[the remark this answers is not in the document]
Comment (lucafedeli88) [reply Source/QED/BreitWheelerEngineWrapper.cpp]: I don't think so: I have to modify the pointer to read throughout all the data

File: Source/QED/BreitWheelerEngineWrapper.cpp [hunk 24]
```diff
@@ -1,16 +1,205 @@
...
+
+    tie(is_ok, innards.ctrl.chi_frac_tpair_how_many, p_data) =
+        parse_raw_data<decltype(innards.ctrl.chi_frac_tpair_how_many)>(
+            p_data, p_last);
+    if(!is_ok) return false;
+
+    //___________________________
+
+    //Data
+    Vector<Real> tndt_coords(innards.ctrl.chi_phot_tdndt_how_many);
```
[the remark this answers is not in the document]
Comment (lucafedeli88) [reply Source/QED/BreitWheelerEngineWrapper.cpp]: That's correct

File: Source/QED/QuantumSyncEngineWrapper.cpp (hunk 6 above)
Comment (lucafedeli88) [reply Source/QED/QuantumSyncEngineWrapper.cpp]: yes. I've created a templated function:
```
    template <class T>
    void add_data_to_vector_char (
        const T* p_data, size_t how_many, amrex::Vector<char>& raw_data)
    {
        raw_data.insert(
            raw_data.end(),
            reinterpret_cast<const char*>(p_data),
            reinterpret_cast<const char*>(p_data) +
            sizeof(T)*how_many
        );
    }
```
and I've moved it to `QedTableParserHelperFunctions.H` .

File: Source/QED/BreitWheelerEngineWrapper.cpp (hunk 3 above)
Comment (lucafedeli88) [reply Source/QED/BreitWheelerEngineWrapper.cpp]: I've removed this lambda and defined a template function in `QedTableParserHelperFunctions.H` (see below)

File: Source/QED/QuantumSyncEngineWrapper.H (hunk 4 above)
Comment (lucafedeli88) [reply Source/QED/QuantumSyncEngineWrapper.H]: see comment for `BreitWheeler`

File: Source/QED/BreitWheelerEngineWrapper.H [hunk 25]
```diff
@@ -2,53 +2,279 @@
...
+    AMREX_GPU_DEVICE
+    AMREX_FORCE_INLINE
+    void operator()(
+    amrex::Real px, amrex::Real py, amrex::Real pz,
+    amrex::Real ex, amrex::Real ey, amrex::Real ez,
+    amrex::Real bx, amrex::Real by, amrex::Real bz,
+    amrex::Real weight,
+    amrex::Real* e_px, amrex::Real* e_py, amrex::Real* e_pz,
+    amrex::Real* p_px, amrex::Real* p_py, amrex::Real* p_pz,
+    amrex::Real* e_weight, amrex::Real* p_weight) const
```
[the remark this answers is not in the document]
Comment (lucafedeli88) [reply Source/QED/BreitWheelerEngineWrapper.H]: Most of them cannot be placed in the constructor, unfortunately: 
- `px,py,pz,ex,ey,ez,bx,by,bz` are particle properties.
- `e_px, e_py, e_pz, p_px, p_py, p_pz` are required for the output of the momenta of the generated particles

`e_weight` and `p_weight` are arrays of length `sampling` with each element equal to `weight/sampling`  . They could be omitted, but the kernel calling the functor would need to perform the same operations. Should I do that?

File: Source/QED/BreitWheelerEngineWrapper.H [hunk 26]
```diff
@@ -2,53 +2,279 @@
...
+            ctrl,
+            rand_zero_one_minus_epsi.data());
+    }
+
+private:
+    //laser wavelenght is not used with SI units
+    const amrex::Real dummy_lambda{1.0};
+
+    const PicsarBreitWheelerCtrl ctrl;
+    const picsar::multi_physics::lookup_2d<amrex::Real> lookup_table;
```
[the remark this answers is not in the document]
Comment (lucafedeli88) [reply Source/QED/BreitWheelerEngineWrapper.H]: See reply above

File: Source/QED/QuantumSyncEngineWrapper.H (hunk 5 above)
Comment (lucafedeli88) [reply Source/QED/QuantumSyncEngineWrapper.H]: see reply above

File: Source/QED/BreitWheelerEngineWrapper.H [hunk 27]
```diff
@@ -2,53 +2,279 @@
...
+
+        return has_event_happend;
+    }
+
+private:
+    //laser wavelenght is not used with SI units
+    const amrex::Real dummy_lambda{1.0};
+
+    const PicsarBreitWheelerCtrl ctrl;
+    const picsar::multi_physics::lookup_1d<amrex::Real> lookup_table;
```
[the remark this answers is not in the document]
Comment (lucafedeli88) [reply Source/QED/BreitWheelerEngineWrapper.H]: I've replaced the lookup table with pointers to the underlying data. At the moment, initializing a PICSAR lookup tables with raw pointers is the safest option to avoid unwanted copies (the constructor of the lookup table instantiates non owning vectors. Therefore, it won't allocate new memory). I am planning to improve the QED PICSAR library in order to have a clearer distinction between classes owning their memory and classes not owning their memory. However, this will take some time, so I won't do that immediately.

File: Source/Particles/MultiParticleContainer.cpp [hunk 28]
```diff
@@ -743,11 +751,209 @@ void MultiParticleContainer::InitQED ()
...
+
+    if(someone_has_breit_wheeler)
+        InitBreitWheeler();
+
+}
+
+void MultiParticleContainer::InitQuantumSync ()
+{
+    bool generate_table;
+    PicsarQuantumSynchrotronCtrl ctrl;
```
[the remark this answers is not in the document]
Comment (lucafedeli88) [reply Source/Particles/MultiParticleContainer.cpp]: Sure, I've added a description

File: Source/QED/BreitWheelerEngineInnards.H (hunk 2 above)
Comment (lucafedeli88) [reply Source/QED/BreitWheelerEngineInnards.H]: sure, I've done that.

File: Source/QED/QuantumSyncEngineWrapper.H (hunk 4 above)
Comment (lucafedeli88) [reply Source/QED/QuantumSyncEngineWrapper.H]: See the comment for the `BreitWheeler` engine

File: Source/QED/QuantumSyncEngineWrapper.H (hunk 5 above)
Comment (lucafedeli88) [reply Source/QED/QuantumSyncEngineWrapper.H]: Yes, I've changed this.

File: Source/QED/BreitWheelerEngineWrapper.cpp (hunk 3 above)
Comment (lucafedeli88) [reply Source/QED/BreitWheelerEngineWrapper.cpp]: Sure. I am using a templated function defined elsewhere however.

File: Source/Particles/MultiParticleContainer.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (MaxThevenet) [reply Source/Particles/MultiParticleContainer.cpp]: Ok agreed, let's keep this.
